system-x/services/kudu/src/main/java/software/tnb/kudu/resource/openshift/RemoteClient.java

```diff
@@ -0,0 +1,192 @@
+import org.awaitility.Awaitility;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.Collectors;
+
+import io.fabric8.kubernetes.client.dsl.ExecListener;
+import io.fabric8.kubernetes.client.dsl.ExecWatch;
+
+/**
+ * Run commands in the client POD using <a href=https://kudu.apache.org/docs/command_line_tools_reference.html>Kudu CLI</a>
+ */
+public class RemoteClient implements KuduClient {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RemoteClient.class);
+
+    private final List<String> masters;
+    private final String clientPodName;
+    private final ObjectMapper mapper;
+    private final OpenshiftClient client;
+
+    private String leader;
+
+    public RemoteClient(List<String> masters, String clientPodName) {
+        this.masters = masters;
+        this.clientPodName = clientPodName;
+        mapper = new ObjectMapper();
+        client = OpenshiftClient.get();
+    }
+
+    @Override
+    public synchronized String findLeaderMasterServer() {
+        if (leader == null) {
+            Awaitility.await("wait for master leader")
+                .atMost(Duration.ofSeconds(20))
+                .ignoreExceptions()
+                .until(() -> {
+                    try (ByteArrayOutputStream out = new ByteArrayOutputStream();
+                         ByteArrayOutputStream err = new ByteArrayOutputStream()) {
+                        ExecWatch watch = runCommand(out, err, "master", "list", "-format=json", getMasters());
+                        handleWatch(watch, err);
+                        if (StringUtils.isBlank(out.toString())) {
+                            throw new RuntimeException("unable to read leader master");
+                        }
+                        List<Map> result = mapper.readValue(out.toString(), List.class);
+                        leader = result.stream().filter(row -> "LEADER".equals(row.get("role")))
+                            .map(row -> row.get("rpc-addresses"))
+                            .map(Objects::toString).findFirst()
+                            .orElseThrow(() -> new IllegalStateException("no leader master found"));
+                        return true;
+                    }
+                });
+        }
+        return leader;
+    }
+
+    @Override
+    public boolean tableExists(String tableName) {
+        try (ByteArrayOutputStream out = new ByteArrayOutputStream();
+             ByteArrayOutputStream err = new ByteArrayOutputStream()) {
```

Comment:
it would be simpler to use `client.podShell(..).execute`  or `executeWithBash`, rather than creating the streams in each method